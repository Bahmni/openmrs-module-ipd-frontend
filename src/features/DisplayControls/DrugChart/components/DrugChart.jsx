import React, { useEffect, useRef } from "react";
import PropTypes from "prop-types";
import useScrollSync from "react-scroll-sync-hook";
import Calendar from "./Calendar";
import CalendarHeader from "./CalendarHeader";
import "../styles/DrugChart.scss";
import DrugList from "./DrugList";
import DrugChartLegend from "../../../../components/AdministrationLegend/AdministrationLegend";
export default function DrugChart(props) {
  const { drugChartData, currentShiftArray, selectedDate } = props;
  const leftPane = useRef(null);
  const rightPane = useRef(null);
  const { registerPane, unregisterPane } = useScrollSync({
    vertical: true,
  });
  const drugChartRowHeight = 66;
  const drugChartHeight = (drugChartData[0].length + 2) * drugChartRowHeight;
  useEffect(() => {
    if (leftPane.current) {
      registerPane(leftPane.current);
    }
    if (rightPane.current) {
      registerPane(rightPane.current);
    }
    return () => {
      if (leftPane.current) {
        unregisterPane(leftPane.current);
      }
      if (rightPane.current) {
        unregisterPane(rightPane.current);
      }
    };
  }, [leftPane, rightPane, registerPane, unregisterPane]);
  return (
    <div className="drug-chart-dashboard">
      <div className="drug-chart" style={{ height: drugChartHeight }}>
        <div
          className="drug-chart-left-panel"
          data-testid="left-panel"
          ref={leftPane}
        >
          <div className="header" />
          <DrugList drugDetails={drugChartData[1]} />
        </div>
        <div
          className="drug-chart-content"
          data-testid="right-panel"
          ref={rightPane}
        >
          <CalendarHeader currentShiftArray={currentShiftArray} />
          <Calendar
            calendarData={drugChartData[0]}
            currentShiftArray={currentShiftArray}
            selectedDate={selectedDate}
          />
        </div>
      </div>
      <DrugChartLegend />
    </div>
  );
}
DrugChart.propTypes = {
  drugChartData: PropTypes.array.isRequired,
  currentShiftArray: PropTypes.array,
  selectedDate: PropTypes.instanceOf(Date),
};
