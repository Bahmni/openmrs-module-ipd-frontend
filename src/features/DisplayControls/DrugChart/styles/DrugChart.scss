.drug-chart-dashboard {
  font-family: "IBM Plex Sans", sans-serif;
}
.drug-chart {
  --time-cell-height: 66px;
  --half-hour-width: 21px;
  --line-width: 1px;

  max-height: calc(8 * var(--time-cell-height));
  display: grid;
  grid-template-columns: 3fr 7fr;
  width: 100%;
  border: 1px solid #ccc;
  margin: 0 auto;
}

.drug-chart-left-panel {
  border-right: 1px solid #ccc;
  font-size: 13px;
  overflow-y: auto;
  height: 100%;
  min-width: 225px;
  box-shadow: 0px 1px 4px 2px rgba(0, 0, 0, 0.25);
  z-index: 20;
  .header {
    height: 63px;
    background-color: #ededed;
    position: sticky;
    top: 0;
  }
  table {
    table-layout: fixed;
    border-spacing: 0;
    width: 100%;
    tbody > tr:nth-child(even) {
      background-color: #f4f4f4;
    }
    tr {
      height: 66px;
      border-bottom: 0;
      td {
        line-height: 20px;
        padding: 0 10px;
        max-width: 100%;
        border: 0;
      }
    }
  }
  -ms-overflow-style: none; /* IE and Edge */
  scrollbar-width: none; /* Firefox */
}

.drug-chart-left-panel::-webkit-scrollbar {
  display: none;
}

.drug-chart-content {
  overflow: auto;
  height: 100%;
}
