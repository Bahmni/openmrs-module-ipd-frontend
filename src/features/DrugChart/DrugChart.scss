.drug-chart-dashboard {
  font-family: "IBM Plex Sans", sans-serif;
}
.drug-chart {
  --time-cell-height: 66px;
  --half-hour-width: 21px;
  --line-width: 1px;

  height: calc(
    7 * var(--time-cell-height)
  ); //TODO: Update the fixed height once the full dashboard is completed
  display: grid;
  grid-template-columns: 3fr 7fr;
  width: 100%;
  border: 1px solid #ccc;
  margin: 0 auto;
}

.drug-chart-left-panel {
  border-right: 1px solid #ccc;
  font-size: 13px;
  overflow-y: auto;
  height: 100%;
  min-width: 225px;
  .header {
    height: 63px;
    background-color: #ededed;
    position: sticky;
    top: 0;
  }
  table {
    table-layout: fixed;
    border-spacing: 0;
    width: 100%;
    tbody > tr:nth-child(even) {
      background-color: #f4f4f4;
    }
    tr {
      height: 66px;
      border-bottom: 0;
      td {
        line-height: 20px;
        padding: 0 10px;
        max-width: 100%;
        border: 0;
      }
    }
  }
}

.drug-chart-content {
  overflow: auto;
  height: 100%;
}
