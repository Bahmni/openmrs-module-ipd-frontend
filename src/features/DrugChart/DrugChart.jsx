import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import Calendar from "../Calendar/Calendar";
import CalendarHeader from "../CalendarHeader/CalendarHeader";
import "./DrugChart.scss";

export const TransformDrugChartData = (drugChartData) => {
  const transformedDrugChartData = drugChartData.map((schedule) => {
    const { slots } = schedule;
    const slotData = {};
    slots.forEach((slot) => {
      const { startDateTime, status } = slot;
      const startDateTimeObj = new Date(startDateTime);
      let adminInfo;
      if (slot.admin) {
        const { administeredBy, administeredAt } = slot.admin;
        const administeredTime = moment(new Date(administeredAt)).format(
          "HH:mm"
        );

        adminInfo = administeredBy + " [" + administeredTime + "]";
      }

      const startHour = startDateTimeObj.getHours();
      const startMinutes = startDateTimeObj.getMinutes();
      slotData[startHour] = {
        minutes: startMinutes,
        status: status,
        administrationInfo: adminInfo,
      };
    });
    return slotData;
  });
  return transformedDrugChartData;
};

export default function DrugChart(props) {
  const { patientId, viewDate = new Date() } = props;
  console.log("DrugChart: patientId: ", patientId);
  console.log("DrugChart: viewDate: ", viewDate);

  const drugChartData = [
    {
      scheduleId: 1,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      comment: "some comment",
      startDate: "2023-08-08T18:30:00.000Z",
      endDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 1,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 11,
          serviceType: "Medication Administration",
          status: "Not-Administered",
          startDateTime: "2023-08-08T08:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
        {
          id: 2,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Administered",
          startDateTime: "2023-08-08T11:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T08:30:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Pending",
          startDateTime: "2023-08-08T14:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
      ],
    },
    {
      scheduleId: 2,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      comment: "some comment",
      scheduleStartDate: "2023-08-08T18:30:00.000Z",
      scheduleEndDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 11,
          serviceType: "Medication Administration",
          status: "Administered-Late",
          startDateTime: "2023-08-08T08:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T08:30:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 4,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Late",
          startDateTime: "2023-08-08T02:30:00.000Z",
          endDateTime: "2023-08-08T03:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T08:30:00.000Z",
            adminid: "1234",
          },
        },
      ],
    },
    {
      scheduleId: 3,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      comment: "some comment",
      scheduleStartDate: "2023-08-08T18:30:00.000Z",
      scheduleEndDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 11,
          serviceType: "Medication Administration",
          status: "Administered",
          startDateTime: "2023-08-08T06:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T06:30:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 4,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Late",
          startDateTime: "2023-08-08T04:30:00.000Z",
          endDateTime: "2023-08-08T05:30:00.000Z",
          notes: "some slot text",
        },
      ],
    },
    {
      scheduleId: 4,
      scheduleUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      scheduleServiceType: "Medication Administration",
      patientUuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
      comment: "some comment",
      startDate: "2023-08-08T18:30:00.000Z",
      endDate: "2023-08-08T18:30:00.000Z",
      slots: [
        {
          id: 1,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 13,
          serviceType: "Medication Administration",
          status: "Not-Administered",
          startDateTime: "2023-08-08T10:45:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
        {
          id: 2,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Administered",
          startDateTime: "2023-08-08T13:35:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
          admin: {
            administeredBy: "Dr. John Doe",
            administeredAt: "2023-08-08T13:40:00.000Z",
            adminid: "1234",
          },
        },
        {
          id: 3,
          uuid: "738aa77d-03fc-438f-a87a-ae8a8867c421",
          orderId: 12,
          serviceType: "Medication Administration",
          status: "Pending",
          startDateTime: "2023-08-08T15:30:00.000Z",
          endDateTime: "2023-08-08T09:30:00.000Z",
          notes: "some slot text",
        },
      ],
    },
  ];
  const transformedDrugchartData = TransformDrugChartData(drugChartData);

  return (
    <div className="drug-chart">
      <div className="drug-chart-left-panel"></div>
      <div className="drug-chart-content">
        <CalendarHeader />
        <Calendar calendarData={transformedDrugchartData} />
      </div>
    </div>
  );
}

DrugChart.propTypes = {
  patientId: PropTypes.string.isRequired,
  viewDate: PropTypes.instanceOf(Date),
};
