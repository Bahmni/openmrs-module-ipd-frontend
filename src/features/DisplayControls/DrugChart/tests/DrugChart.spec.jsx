import React from "react";
import { render } from "@testing-library/react";
import DrugChart from "../components/DrugChart";
import { drugChartData } from "./DrugChartMockData";

const mockCalendar = jest.fn();
jest.mock("../components/CalendarHeader", () => {
  return function CalendarHeader() {
    return <div>CalendarHeader</div>;
  };
});

jest.mock("../components/Calendar", () => {
  return function Calendar(props) {
    mockCalendar(props);
    return <div>Calendar</div>;
  };
});

const mockRegisterPane = jest.fn();
const mockUnregisterPane = jest.fn();

jest.mock("react-scroll-sync-hook", () => {
  return function useScrollSync() {
    return {
      registerPane: mockRegisterPane,
      unregisterPane: mockUnregisterPane,
    };
  };
});

const MockTooltipCarbon = jest.fn();

jest.mock("bahmni-carbon-ui", () => {
  return {
    TooltipCarbon: (props) => {
      MockTooltipCarbon(props);
      return <div>TooltipCarbon</div>;
    },
  };
});

describe("DrugChart", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(<DrugChart drugChartData={drugChartData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call register pane when ref is on rightpane", () => {
    const { getByTestId } = render(<DrugChart drugChartData={drugChartData} />);
    const rightPanel = getByTestId("right-panel");
    expect(mockRegisterPane).toHaveBeenCalledWith(rightPanel);
  });

  it("should  unregisterPane when component is  unmounted", () => {
    const { unmount } = render(<DrugChart drugChartData={drugChartData} />);
    unmount();
    expect(mockUnregisterPane).toHaveBeenCalled();
  });

  it("should synchronize scroll position of the right panel and the left panel when scrolled to bottom", () => {
    const { getByTestId } = render(<DrugChart drugChartData={drugChartData} />);
    const leftPanel = getByTestId("left-panel");
    const rightPanel = getByTestId("right-panel");
    leftPanel.scrollBottom = 100;
    rightPanel.scrollBottom = 100;
    expect(leftPanel.scrollBottom).toEqual(rightPanel.scrollBottom);
  });
});
