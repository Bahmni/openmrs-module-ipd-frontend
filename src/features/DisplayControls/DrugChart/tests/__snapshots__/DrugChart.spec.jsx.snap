// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugChart should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="drug-chart-dashboard"
  >
    <div
      class="drug-chart"
      style="height: 264px;"
    >
      <div
        class="drug-chart-left-panel"
        data-testid="left-panel"
      >
        <div
          class="header"
        />
        <div>
          <table
            style="overflow: hidden;"
          >
            <tbody>
              <tr>
                <td>
                  <div
                    class="drug-name-container"
                  >
                    <div
                      class="drug-chart-drug-name"
                    >
                      Suxamethonium Chloride 100 mg/2 mL Ampoule (50 mg/mL )
                    </div>
                    <div>
                      TooltipCarbon
                    </div>
                    <div
                      class="drug-name-cell"
                    >
                      <div
                        class="bx--tag bx--tag--blue"
                        id="tag-1"
                      >
                        <span
                          title="Rx"
                        >
                          Rx
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="dosage"
                  >
                    <span>
                      2
                    </span>
                    <span>
                       - mg
                    </span>
                    <span>
                       - Intravenous
                    </span>
                    <span>
                       - 4 Day(s)
                    </span>
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  <div
                    class="drug-name-container"
                  >
                    <div
                      class="drug-chart-drug-name"
                    >
                      Enalapril 5 mg Tablet
                    </div>
                    <div>
                      TooltipCarbon
                    </div>
                    <div
                      class="drug-name-cell"
                    >
                      <div
                        class="bx--tag bx--tag--blue"
                        id="tag-2"
                      >
                        <span
                          title="Rx"
                        >
                          Rx
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="dosage"
                  >
                    <span>
                      1
                    </span>
                    <span>
                       - Tablet(s)
                    </span>
                    <span>
                       - Oral
                    </span>
                    <span>
                       - 4 Day(s)
                    </span>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="drug-chart-content"
        data-testid="right-panel"
      >
        <div>
          CalendarHeader
        </div>
        <div>
          Calendar
        </div>
      </div>
    </div>
    <div
      class="drug-chart-legend"
    >
      <div>
        <test-file-stub />
        Pending
      </div>
      <div>
        <test-file-stub />
        Late
      </div>
      <div>
        <test-file-stub />
        Administered
      </div>
      <div>
        <test-file-stub />
        Administered Late
      </div>
      <div>
        <test-file-stub />
        Not Administered
      </div>
    </div>
  </div>
</DocumentFragment>
`;
